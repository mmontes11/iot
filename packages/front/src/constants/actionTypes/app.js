export const TOGGLE_HAMBURGER_MENU = "TOGGLE_HAMBURGER_MENU";
export const TOGGLE_MAP_DIALOG = "TOGGLE_MAP_DIALOG";
export const SET_SHOW_ERROR = "SET_SHOW_ERROR";
