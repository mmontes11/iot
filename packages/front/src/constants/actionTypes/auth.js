export const IS_AUTH = "IS_AUTH";
export const SET_USERNAME = "SET_USERNAME";
export const SET_PASSWORD = "SET_PASSWORD";
export const LOGIN_REQUEST = "LOGIN_REQUEST";
export const LOGIN_REQUEST_SUCCESS = "LOGIN_REQUEST_SUCCESS";
export const LOGIN_REQUEST_ERROR = "LOGIN_REQUEST_ERROR";
