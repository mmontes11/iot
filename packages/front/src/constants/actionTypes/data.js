export const DATA_UPDATED = "DATA_UPDATED";
export const DATA_REQUEST = "DATA_REQUEST";
export const DATA_REQUEST_SUCCESS = "DATA_REQUEST_SUCCESS";
export const DATA_REQUEST_ERROR = "DATA_REQUEST_ERROR";
export const ADD_DATA_ITEM = "ADD_DATA_ITEM";
export const RESET_DATA = "RESET_DATA";
