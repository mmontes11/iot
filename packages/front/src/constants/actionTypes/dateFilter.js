export const DATE_FILTER_SELECTOR_TOGGLE = "DATE_FILTER_SELECTOR_TOGGLE";
export const DATE_FILTER_TIME_PERIOD_SELECT = "DATE_FILTER_TIME_PERIOD_SELECT";
export const DATE_FILTER_TIME_PERIOD_UPDATED = "DATE_FILTER_TIME_PERIOD_UPDATED";
export const DATE_FILTER_TIME_PERIODS_REQUEST = "DATE_FILTER_TIME_PERIODS_REQUEST";
export const DATE_FILTER_TIME_PERIODS_REQUEST_SUCCESS = "DATE_FILTER_TIME_PERIODS_REQUEST_SUCCESS";
export const DATE_FILTER_TIME_PERIODS_REQUEST_ERROR = "DATE_FILTER_TIME_PERIODS_REQUEST_ERROR";
export const DATE_FILTER_TIME_PERIODS_UPDATED = "DATE_FILTER_TIME_PERIODS_UPDATED";
export const DATE_FILTER_START_DATE_UPDATED = "DATE_FILTER_START_DATE_UPDATED";
export const DATE_FILTER_END_DATE_UPDATED = "DATE_FILTER_END_DATE_UPDATED";
