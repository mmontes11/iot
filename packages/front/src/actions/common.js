import { RESET } from "constants/actionTypes/common";

export const reset = preserveError => dispatch => {
  dispatch({ type: RESET, preserveError });
};
