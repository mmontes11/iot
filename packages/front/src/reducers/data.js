import { RESET } from "constants/actionTypes/common";
import {
  DATA_REQUEST,
  DATA_REQUEST_SUCCESS,
  DATA_REQUEST_ERROR,
  DATA_UPDATED,
  ADD_DATA_ITEM,
  RESET_DATA,
} from "constants/actionTypes/data";

export const initialState = {
  items: [],
  things: [],
  isLoading: false,
};

export default (state = initialState, { type, items, things, item }) => {
  switch (type) {
    case DATA_REQUEST:
      return { ...state, isLoading: true };
    case DATA_REQUEST_SUCCESS:
      return { ...state, isLoading: false };
    case DATA_REQUEST_ERROR:
      return { ...state, isLoading: false };
    case DATA_UPDATED:
      return { ...state, items, things };
    case ADD_DATA_ITEM:
      return { ...state, items: [...state.items, item] };
    case RESET_DATA:
      return initialState;
    case RESET:
      return initialState;
    default:
      return state;
  }
};
