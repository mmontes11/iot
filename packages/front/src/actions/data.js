import {
  DATA_UPDATED,
  DATA_REQUEST,
  DATA_REQUEST_SUCCESS,
  DATA_REQUEST_ERROR,
  ADD_DATA_ITEM,
  RESET_DATA,
} from "constants/actionTypes/data";
import { EVENT_TYPE, MEASUREMENT_TYPE } from "constants/observationTypes";
import iotClient from "lib/iotClient";
import { SocketController } from "helpers/socketController";
import { THING_FILTER_TYPE, DATE_FILTER_TYPE } from "constants/filterTypes";
import { RESET } from "constants/actionTypes/common";
import * as fromState from "reducers";
import { TYPE, OBSERVATION, GROUPBY, THING } from "constants/params";

const getParams = (filterItems, observation, groupBy, thing, timePeriod, startDate, endDate) => {
  let params = {
    type: observation,
  };
  if (groupBy) {
    params = {
      ...params,
      groupBy,
    };
  }
  if (filterItems.includes(THING_FILTER_TYPE) && thing) {
    params = {
      ...params,
      thing,
    };
  }
  if (filterItems.includes(DATE_FILTER_TYPE)) {
    if (timePeriod) {
      params = {
        ...params,
        timePeriod,
      };
    } else {
      if (startDate) {
        params = {
          ...params,
          startDate: startDate.toISOString(),
        };
      }
      if (endDate) {
        params = {
          ...params,
          endDate: endDate.toISOString(),
        };
      }
    }
  }
  return params;
};

const requestStats = (
  filterItems,
  type,
  observation,
  thing,
  timePeriod,
  startDate,
  endDate,
  onStart,
  onSuccess,
  onError,
) => {
  const params = getParams(filterItems, observation, null, thing, timePeriod, startDate, endDate);
  if (type === EVENT_TYPE) {
    onStart();
    iotClient.eventService
      .getStats(params)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else if (type === MEASUREMENT_TYPE) {
    onStart();
    iotClient.measurementService
      .getStats(params)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else {
    onError(new Error("Invalid type"));
  }
};

export const getStats = () => (dispatch, getState) => {
  const state = getState();
  const filterItems = state.filters.items;
  const type = fromState.getParam(state, TYPE).selectedItem;
  const observation = fromState.getParam(state, OBSERVATION).selectedItem;
  const thing = state.filters.thingFilter.selectedItem;
  const {
    timePeriod: { selectedItem: selectedTimePeriod },
    custom: { startDate, endDate },
  } = state.filters.dateFilter;
  if (!type || !observation) {
    return;
  }
  requestStats(
    filterItems,
    type,
    observation,
    thing,
    selectedTimePeriod,
    startDate,
    endDate,
    () => dispatch({ type: DATA_REQUEST }),
    res => {
      dispatch({ type: DATA_REQUEST_SUCCESS });
      dispatch({ type: DATA_UPDATED, items: res.body.stats });
    },
    () => {
      dispatch({ type: DATA_REQUEST_ERROR });
      dispatch({ type: RESET, preserveError: true });
    },
  );
};

const requestData = (
  filterItems,
  type,
  observation,
  groupBy,
  thing,
  timePeriod,
  startDate,
  endDate,
  onStart,
  onSuccess,
  onError,
) => {
  const params = getParams(filterItems, observation, groupBy, thing, timePeriod, startDate, endDate);
  if (type === EVENT_TYPE) {
    onStart();
    iotClient.eventService
      .getData(params)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else if (type === MEASUREMENT_TYPE) {
    onStart();
    iotClient.measurementService
      .getData(params)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else {
    onError(new Error("Invalid type"));
  }
};

export const getData = () => (dispatch, getState) => {
  const state = getState();
  const filterItems = state.filters.items;
  const type = fromState.getParam(state, TYPE).selectedItem;
  const observation = fromState.getParam(state, OBSERVATION).selectedItem;
  const groupBy = fromState.getParam(state, GROUPBY).selectedItem;
  const thing = state.filters.thingFilter.selectedItem;
  const {
    timePeriod: { selectedItem: selectedTimePeriod },
    custom: { startDate, endDate },
  } = state.filters.dateFilter;
  if (!type || !observation) {
    return;
  }
  requestData(
    filterItems,
    type,
    observation,
    groupBy,
    thing,
    selectedTimePeriod,
    startDate,
    endDate,
    () => dispatch({ type: DATA_REQUEST }),
    ({ body: { data, things } }) => {
      dispatch({ type: DATA_REQUEST_SUCCESS });
      dispatch({ type: DATA_UPDATED, items: data, things });
    },
    () => {
      dispatch({ type: DATA_REQUEST_ERROR });
      dispatch({ type: RESET, preserveError: true });
    },
  );
};

let socketController = null;

export const finishRealTimeData = () => () => {
  if (socketController !== null) {
    socketController.close();
  }
};

export const startRealTimeData = () => (dispatch, getState) => {
  const state = getState();
  const thing = fromState.getParam(state, THING).selectedItem;
  const type = fromState.getParam(state, TYPE).selectedItem;
  if (!thing || !type) {
    return;
  }
  finishRealTimeData()();
  const onData = data => {
    const {
      data: { items },
    } = getState();
    if (items.length === 0) {
      dispatch({ type: DATA_REQUEST_SUCCESS });
    }
    dispatch({ type: ADD_DATA_ITEM, item: data });
  };
  const onError = error => {
    dispatch({ type: DATA_REQUEST_ERROR, error });
    dispatch({ type: RESET, preserveError: true });
  };
  socketController = new SocketController(thing, type, onData, onError);
  socketController.listen();
  dispatch({ type: DATA_REQUEST });
};

export const resetData = () => dispatch => {
  dispatch({ type: RESET_DATA });
};
