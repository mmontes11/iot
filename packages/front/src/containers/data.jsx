import React from "react";
import PropTypes from "prop-types";
import { handleDataParams } from "hocs/dataParams";
import store from "config/store";
import { getData } from "actions/data";
import { reset } from "actions/common";
import DataParamsPanel from "containers/dataParamsPanel";
import FiltersPanelContainer from "containers/filtersPanelContainer";
import Charts from "containers/charts";
import { LINECHART } from "constants/chartTypes";
import { TYPE, OBSERVATION, GROUPBY } from "constants/params";

const Data = ({ onParamsSelected, onFiltersSelected, onReset }) => (
  <div className="container is-fluid section">
    <div className="columns">
      <div className="column is-three-quarters">
        <DataParamsPanel onParamsSelected={onParamsSelected} onReset={onReset} />
        <Charts chartType={LINECHART} />
      </div>
      <div className="column is-one-quarter">
        <FiltersPanelContainer onFiltersSelected={onFiltersSelected} />
      </div>
    </div>
  </div>
);

Data.propTypes = {
  onParamsSelected: PropTypes.func.isRequired,
  onFiltersSelected: PropTypes.func.isRequired,
  onReset: PropTypes.func.isRequired,
};

const withDataParams = handleDataParams(
  "data",
  [TYPE, OBSERVATION, GROUPBY],
  () => store.dispatch(getData()),
  () => store.dispatch(reset()),
);

export default withDataParams(Data);
