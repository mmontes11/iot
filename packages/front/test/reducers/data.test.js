import deepFreeze from "deep-freeze";
import dataReducer from "reducers/data";
import { RESET } from "constants/actionTypes/common";
import { DATA_REQUEST, DATA_REQUEST_SUCCESS, DATA_REQUEST_ERROR, DATA_UPDATED } from "constants/actionTypes/data";
import { initialState } from "../constants/index";

const dataInitialState = initialState.data;
deepFreeze(dataInitialState);

describe("reducers/data", () => {
  it("has a default state", () => {
    expect(dataReducer(undefined, { type: "@init" })).toMatchSnapshot();
  });
  it("reduces nothing", () => {
    expect(dataReducer(dataInitialState, { type: "WHATEVER" })).toMatchSnapshot();
  });
  it("reduces DATA_REQUEST", () => {
    expect(dataReducer(dataInitialState, { type: DATA_REQUEST })).toMatchSnapshot();
  });
  it("reduces DATA_REQUEST_SUCCESS", () => {
    expect(dataReducer(dataInitialState, { type: DATA_REQUEST_SUCCESS })).toMatchSnapshot();
  });
  it("reduces DATA_REQUEST_ERROR", () => {
    expect(dataReducer(dataInitialState, { type: DATA_REQUEST_ERROR })).toMatchSnapshot();
  });
  it("reduces DATA_UPDATED", () => {
    expect(dataReducer(dataInitialState, { type: DATA_UPDATED, items: [] })).toMatchSnapshot();
  });
  it("reduces RESET", () => {
    expect(dataReducer(dataInitialState, { type: RESET })).toMatchSnapshot();
  });
});
