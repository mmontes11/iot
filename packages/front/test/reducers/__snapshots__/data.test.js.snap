// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/data has a default state 1`] = `
Object {
  "isLoading": false,
  "items": Array [],
  "things": Array [],
}
`;

exports[`reducers/data reduces DATA_REQUEST 1`] = `
Object {
  "isLoading": true,
  "items": Array [],
  "things": Array [],
}
`;

exports[`reducers/data reduces DATA_REQUEST_ERROR 1`] = `
Object {
  "isLoading": false,
  "items": Array [],
  "things": Array [],
}
`;

exports[`reducers/data reduces DATA_REQUEST_SUCCESS 1`] = `
Object {
  "isLoading": false,
  "items": Array [],
  "things": Array [],
}
`;

exports[`reducers/data reduces DATA_UPDATED 1`] = `
Object {
  "isLoading": false,
  "items": Array [],
  "things": undefined,
}
`;

exports[`reducers/data reduces RESET 1`] = `
Object {
  "isLoading": false,
  "items": Array [],
  "things": Array [],
}
`;

exports[`reducers/data reduces nothing 1`] = `
Object {
  "isLoading": false,
  "items": Array [],
  "things": Array [],
}
`;
